web/frontend/src/components/gamepad.vue

```diff
@@ -37,131 +37,188 @@ const curStates = $ref<Record<string, number>>({
-    if ((/X|Y|Z$/u).test(key)) {
-      newEvent.setControl(`Absolute${key}`);
-      newEvent.setEvent('PositionChangeAbs');
-    } else {
-      newEvent.setControl(`Button${key}`);
-      newEvent.setEvent(
-        value ? 'ButtonPress' : 'ButtonRelease'
-      );
-    }
+  const nowTS = nextTS();
 
-    if (Number.isNaN(value)) {
-      newEvent.setEvent('Disconnect');
-      newEvent.setValue(0);
-    } else {
-      newEvent.setValue(value);
+  try {
+    for (const [key, value] of Object.entries(curStates)) {
+      if (value === prevStates[key] || (Number.isNaN(value) && Number.isNaN(prevStates[key]))) {
+        continue;
+      }
+      const newEvent = new InputController.Event();
+      nowTS.setNanos(nowTS.getNanos() + 1);
+      newEvent.setTime(nowTS);
+      if ((/X|Y|Z$/u).test(key)) {
+        newEvent.setControl(`Absolute${key}`);
+        newEvent.setEvent('PositionChangeAbs');
+      } else {
+        newEvent.setControl(`Button${key}`);
+        newEvent.setEvent(value ? 'ButtonPress' : 'ButtonRelease');
+      }
+
+      if (Number.isNaN(value)) {
+        newEvent.setEvent('Disconnect');
+        newEvent.setValue(0);
+      } else {
+        newEvent.setValue(value);
+      }
+      sendEvent(newEvent);
     }
-    sendEvent(newEvent);
+  } finally {
+    lastTS = nowTS;
   }
 };
 
 const tick = () => {
-  let gamepadFound = false;
+  if (!gamepad) {
+    return;
+  }
+
+  prevStates = { ...prevStates, ...curStates };
+
+  // eslint-disable-next-line unicorn/no-unsafe-regex
+  const re = /^-?\d+(?:.\d{0,4})?/u;
+  const trunc = (val: number): number => {
+    if (Number.isNaN(val)) {
+      return val;
+    }
+    const match = val.toString().match(re);
+    if (match && match.length === 0) {
+      return val;
+    }
+    return Number(match![0]!);
+  };
+
+  curStates.X = trunc(gamepad.axes[0]!);
+  curStates.Y = trunc(gamepad.axes[1]!);
+  curStates.RX = trunc(gamepad.axes[2]!);
+  curStates.RY = trunc(gamepad.axes[3]!);
+  curStates.Z = trunc(gamepad.buttons[6]!.value);
+  curStates.RZ = trunc(gamepad.buttons[7]!.value);
+  curStates.Hat0X = trunc((gamepad.buttons[14]!.value * -1) + gamepad.buttons[15]!.value);
+  curStates.Hat0Y = trunc((gamepad.buttons[12]!.value * -1) + gamepad.buttons[13]!.value);
+  curStates.South = trunc(gamepad.buttons[0]!.value);
+  curStates.East = trunc(gamepad.buttons[1]!.value);
+  curStates.West = trunc(gamepad.buttons[2]!.value);
+  curStates.North = trunc(gamepad.buttons[3]!.value);
+  curStates.LT = trunc(gamepad.buttons[4]!.value);
+  curStates.RT = trunc(gamepad.buttons[5]!.value);
+  curStates.Select = trunc(gamepad.buttons[8]!.value);
+  curStates.Start = trunc(gamepad.buttons[9]!.value);
+  curStates.LThumb = trunc(gamepad.buttons[10]!.value);
+  curStates.RThumb = trunc(gamepad.buttons[11]!.value);
+  curStates.Menu = trunc(gamepad.buttons[16]!.value);
+
+  if (enabled) {
+    processEvents();
+  }
+  handle = window.setTimeout(tick, 10);
```

Comment:
previously we were using request animation frame which is variable rate and on a macbook like mine, is well over 60fps. This is simply too much to send for unary via a browser. In the future, this could be more okay (maybe), if we used a streaming api instead. Especially for grpc-web where each unary is a request... a bidi/client streaming, when supported, can pevent network tab from bogging down

Reply:
Yeah this makes sense 👍 RAF is not a good fit here. Curious: do we send time deltas for each of these ticks? Wondering because of how setTimeout (and RAF) only guarantee minimum times until the next callback.

Reply:
we do not but don't need it since the remote control doesn't require it